src/components/PupilComponents/LessonEngineProvider/LessonEngineProvider.tsx

```diff
@@ -9,34 +16,127 @@ export const lessonSections = [
   "review",
 ] as const;
 
-export const lessonReviewSections = [
+export const allLessonReviewSections = [
   "intro",
   "starter-quiz",
   "video",
   "exit-quiz",
 ] as const;
 
 export type LessonSection = (typeof lessonSections)[number];
+export type LessonReviewSection = (typeof allLessonReviewSections)[number];
 
 export const isLessonSection = (
   currentSection: string,
 ): currentSection is LessonSection => {
   return lessonSections.includes(currentSection as LessonSection);
 };
 
-export type LessonSectionResult = { grade: number; numQuestions: number };
+export function isLessonReviewSection(
+  section: string,
+): section is LessonReviewSection {
+  return lessonSections.includes(section as LessonReviewSection);
+}
+
+type QuizResult = { grade: number; numQuestions: number };
+type LessonSectionState = {
+  isComplete: boolean;
+} & Partial<QuizResult>;
+type LessonEngineAction =
+  | {
+      type: "setCurrentSection";
+      section: LessonSection;
+    }
+  | {
+      type: "completeSection";
+      section: LessonReviewSection;
+    }
+  | {
+      type: "updateQuizResult";
+      result: QuizResult;
+    }
+  | {
+      type: "proceedToNextSection";
+    };
+
+type LessonEngineState = {
+  currentSection: LessonSection;
+  lessonReviewSections: Readonly<LessonReviewSection[]>;
+  sections: Partial<Record<LessonReviewSection, LessonSectionState>>;
+};
 
-type LessonSectionResults = Partial<Record<LessonSection, LessonSectionResult>>;
+const lessonEngineReducer: Reducer<LessonEngineState, LessonEngineAction> = (
+  currentState,
+  action,
+) => {
+  switch (action.type) {
+    case "setCurrentSection":
+      return { ...currentState, currentSection: action.section };
+    case "completeSection": {
+      const newState = {
+        ...currentState,
+        sections: {
+          ...currentState.sections,
+          [action.section]: {
+            ...currentState.sections[action.section],
+            isComplete: true,
+          },
+        },
+      };
+
+      // redirect the user according to what sections have been completed
+      newState.currentSection = currentState.lessonReviewSections.every(
+        (section) => newState.sections[section]?.isComplete,
+      )
+        ? "review"
+        : "overview";
+
+      return newState;
+    }
+    case "proceedToNextSection": {
+      const nextSection =
+        currentState.lessonReviewSections.find(
+          (section) => !currentState.sections[section]?.isComplete,
+        ) ?? currentState.lessonReviewSections[0];
+
+      if (!nextSection) {
+        throw new Error("Lesson contains no sections to proceed to");
+      }
+
+      return { ...currentState, currentSection: nextSection };
+    }
+    case "updateQuizResult": {
+      if (!isLessonReviewSection(currentState.currentSection)) {
+        throw new Error(
+          `Cannot update quiz result for non-review section '${currentState.currentSection}'`,
+        );
+      }
```

Comment:
When I tightened things up by only allowing review sections to have quiz results this little corner-case appeared (it could be encountered if the `updateQuizResult` action were triggered on the `overview` page) It's not possible to reach this currently in the product — but this will alert a dev should a subtle bug be introduced in the future.